### PR TITLE
Bug fix/internal issue #600 - Late document materialization

### PR DESCRIPTION
**Pull request by Dronplane**
This PR introduces new optimizer rule  "late-document-materialization"
This rule adds ability to delay actual reading of document from storage engine(materialization) as long as possible, if there is prove that not all documents will be returned. This is determined by presense of SortNode and LimitNode in the loop. As without SortNode  there will be just skiprows and nothing to optimize. Without limit node all selected documents will be returned, so no reason to delay materialization.
Materialization is performed on new MaterializeNode while View node emits only documents ids. If there are several sort+limit node pairs, lowest node is selected  to be appended with materializer (while tracking usage of actual document - it should not be used before that or at that pair). On cluster materialization only could be delayed before first RemoteNode (as network hops for each document later considered more expensive than just  getting them all at the beginning).
Added unit and js tests.

### CODE REVIEW

**Comment by Dronplane** [comment]
https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/6265/

**Comment by Dronplane** [comment]
https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/6274/

**Comment by Dronplane** [comment]
https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/6283/

File: arangod/Aql/SortExecutor.cpp [hunk 1]
```diff
@@ -179,3 +228,59 @@ std::pair<ExecutionState, size_t> SortExecutor::expectedNumberOfRows(size_t atMo
...
+      TRI_ASSERT(ctx._inputRow->isInitialized());
+      TRI_ASSERT(ctx._infos);
+      arangodb::aql::AqlValue a{ arangodb::aql::AqlValueHintCopy(doc.begin()) };
+      bool mustDestroy = true;
+      arangodb::aql::AqlValueGuard guard{ a, mustDestroy };
+      ctx._outputRow->moveValueInto(ctx._infos->outputMaterializedDocumentRegId(), *ctx._inputRow, guard);
+    };
+  },
+
+    [](ReadContext& ctx) {
```
Comment (graetzer) [inline arangod/Aql/SortExecutor.cpp]: you could write this stuff as a standalone function and I think it would be more readable

**Comment by Dronplane** [comment]
https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/6310/

**Comment by Dronplane** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6559/

File: arangod/Aql/RegisterPlan.cpp [hunk 2]
```diff
@@ -238,6 +238,7 @@ void RegisterPlan::after(ExecutionNode* en) {
     case ExecutionNode::SINGLETON:
     case ExecutionNode::FILTER:
     case ExecutionNode::LIMIT:
+    case ExecutionNode::MATERIALIZE:
```
Comment (goedderz) [inline arangod/Aql/RegisterPlan.cpp]: The MaterializeNode should add the output variable of the document, not the view.

File: arangod/Aql/IResearchViewNode.cpp [hunk 3]
```diff
@@ -640,6 +640,21 @@ SnapshotPtr snapshotSingleServer(IResearchViewNode const& node, transaction::Met
...
+const char* NODE_OUT_VARIABLE_PARAM = "outVariable";
+const char* NODE_OUT_NM_DOC_PARAM = "outNmDocId";
+const char* NODE_OUT_NM_COL_PARAM = "outNmColPtr";
+const char* NODE_CONDITION_PARAM = "condition";
+const char* NODE_SCORERS_PARAM = "scorers";
+const char* NODE_SHARDS_PARAM = "shards";
+const char* NODE_OPTIONS_PARAM = "options";
+const char* NODE_VOLATILITY_PARAM = "volatility";
+const char* NODE_PRIMARY_SORT_PARAM = "primarySort";
+const char* NODE_PRIMARY_SORT_BUCKETS_PARAM = "primarySortBuckets";
```
[the remark this answers is not in the document]
Comment (Dronplane) [reply arangod/Aql/IResearchViewNode.cpp]: These are internal business of IResearchViewNode. I think it is better to keep them locally as this is not intended to be used outside ViewNode class

File: arangod/Aql/IResearchViewNode.cpp [hunk 4]
```diff
@@ -858,15 +858,13 @@ void IResearchViewNode::planNodeRegisters(
...
   // create a copy of the last value here
   // this is requried because back returns a reference and emplace/push_back
   // may invalidate all references
-  aql::RegisterId const registerId = 1 + nrRegs.back();
-  nrRegs.emplace_back(registerId);
-
-  varInfo.emplace(_outVariable->id, aql::VarInfo(depth, totalNrRegs++));
-
+  aql::RegisterCount const prevRegistersCount = nrRegs.back();
+  nrRegs.emplace_back(prevRegistersCount);
```
Comment (goedderz) [inline arangod/Aql/IResearchViewNode.cpp]: Is this correct (omitting the +1), or does it essentially have to be `prevRegistersCount + nrRegsHere.back()`?

File: arangod/Aql/RegisterPlan.cpp [hunk 5]
```diff
@@ -315,7 +314,20 @@ void RegisterPlan::after(ExecutionNode* en) {
...
+    case ExecutionNode::MATERIALIZE: {
+      depth++;
+      nrRegsHere.emplace_back(1);
+      // create a copy of the last value here
+      // this is required because back returns a reference and emplace/push_back
+      // may invalidate all references
+      RegisterId registerId = nrRegs.back() + 1;
+      nrRegs.emplace_back(registerId);
+      auto ep = ExecutionNode::castTo<MaterializeNode const*>(en);
+      TRI_ASSERT(ep != nullptr);
```
Comment (goedderz) [inline arangod/Aql/RegisterPlan.cpp]: `castTo` already asserts that.

**Pull request merged**